### PR TITLE
Enabling default arguments by default.

### PR DESCRIPTION
**Pull request by aneeshdk**
Enabling default arguments by default.

This changelist enables the default arguments scenario which does not
require split scope by default. For split scope scenarios a new flag is
added which is disabled by default. Also if a split scope required
scenario occurs in a normal mode a syntax error is thrown. That includes
scenarios where,
1. A function definition occurs in the formal parameters which refer to
one of the formals
2. Eval occurs in formals.

Fixed couple of bugs. Moved the split scope test cases to a separate file.

### CODE REVIEW

**Comment by aneeshdk** [comment]
@akroshg @abchatra Can you please take a look?

File: Lib/Runtime/ByteCode/FuncInfo.cpp [hunk 1]
```diff
@@ -435,14 +435,42 @@ void FuncInfo::OnStartVisitScope(Scope *scope)
...
-        // If the scopes are unmerged and we are visiting the parameter scope, the child scope will be the function body scope.
-        Assert(!scope->GetCanMergeWithBodyScope() || this->GetCurrentChildScope()->GetEnclosingScope() == scope || this->GetCurrentChildScope() == nullptr);
-    }
-    else
-    {
-        Assert(this->GetCurrentChildScope() == scope->GetEnclosingScope() || this->GetCurrentChildScope() == nullptr);
+        if (scope->GetScopeType() == ScopeType_Parameter)
+        {
+            // If the scopes are unmerged and we are visiting the parameter scope, the child scope will be the function body scope.  
+            Assert(!scope->GetCanMergeWithBodyScope() || this->GetCurrentChildScope()->GetEnclosingScope() == scope);
```
Comment (akroshg) [inline Lib/Runtime/ByteCode/FuncInfo.cpp]: use childScope instead of this->GetCurrentChildScope()

File: Lib/Common/ConfigFlagsList.h [hunk 2]
```diff
@@ -857,7 +858,8 @@ FLAGPR           (Boolean, ES6, ES6Species             , "Enable ES6 '@@species'
 FLAGPR_REGOVR_EXP(Boolean, ES6, ES7AsyncAwait          , "Enable ES7 'async' and 'await' keywords"                  , DEFAULT_CONFIG_ES6AsyncAwait)
 FLAGPR           (Boolean, ES6, ES6Classes             , "Enable ES6 'class' and 'extends' keywords"                , DEFAULT_CONFIG_ES6Classes)
 FLAGPR           (Boolean, ES6, ES6DateParseFix        , "Enable ES6 Date.parse fixes"                              , DEFAULT_CONFIG_ES6DateParseFix)
-FLAGPR_REGOVR_EXP(Boolean, ES6, ES6DefaultArgs         , "Enable ES6 Default Arguments"                             , DEFAULT_CONFIG_ES6DefaultArgs)
+FLAGPR           (Boolean, ES6, ES6DefaultArgs         , "Enable ES6 Default Arguments"                             , DEFAULT_CONFIG_ES6DefaultArgs)
+FLAGPR_REGOVR_EXP(Boolean, ES6, ES6DefaultArgsSplitScope, "Enable ES6 Default Arguments in split scope scenarios"   , DEFAULT_CONFIG_ES6DefaultArgsSplitScope)
```
Comment (akroshg) [inline Lib/Common/ConfigFlagsList.h]: Would it be better to say 'Enable parameters to have its own scope' ?

File: Lib/Parser/perrors.h [hunk 3]
```diff
@@ -92,3 +92,5 @@ LSC_ERROR_MSG(1082, ERRNonSimpleParamListInStrictMode, "Cannot apply strict mode
 LSC_ERROR_MSG(1083, ERRBadAwait, "'await' expression not allowed in this context")
 
 LSC_ERROR_MSG(1084, ERRGetterMustHaveNoParameters, "Getter functions must have no parameters")
+LSC_ERROR_MSG(1085, ERRFuncRefFormalNotSupportedInParamScope, "Formals cannot contain functions definitions that reference them")
+LSC_ERROR_MSG(1086, ERREvalNotSupportedInParamScope, "Non-simple parameter list cannot have eval")
```
Comment (akroshg) [inline Lib/Parser/perrors.h]: How about - "'eval' is not allowed in the default initializer" ?

File: test/es6/rlexe.xml [hunk 4]
```diff
@@ -682,6 +682,12 @@
   </test>
   <test>
     <default>
+      <files>default-splitscope.js</files>
+      <compile-flags>-force:CachedScope -ES6DefaultArgsSplitScope -ES6Generators -args summary -endargs</compile-flags>
```
Comment (akroshg) [inline test/es6/rlexe.xml]: Why is this -force:CachedScope

**Comment by akroshg** [comment]
Apart from comments above looks good to me.

File: Lib/Runtime/ByteCode/FuncInfo.cpp (hunk 1 above)
Comment (aneeshdk) [inline Lib/Runtime/ByteCode/FuncInfo.cpp]: Done

File: Lib/Common/ConfigFlagsList.h (hunk 2 above)
Comment (aneeshdk) [inline Lib/Common/ConfigFlagsList.h]: Changed it

File: Lib/Parser/perrors.h (hunk 3 above)
Comment (aneeshdk) [inline Lib/Parser/perrors.h]: Done

**Comment by dilijev** [comment]
@dotnet-bot test EOL Check please

**Pull request merged**